### PR TITLE
HHH-17783, HHH-17918 - Allow early initialization of entity instance by non-owning initializer

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-17783
https://hibernate.atlassian.net/browse/HHH-17918

Backport of #8149

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java [hunk 1]
```diff
@@ -215,23 +216,7 @@ public Object forceInitialize(Object target, String attributeName, SharedSession
...
-					),
-					entityKey,
-					persister.getVersion( target ),
-					LockMode.NONE,
-					// we assume an entry exists in the db
-					true,
-					persister,
-					true
-			);
+			session.getPersistenceContext().addEnhancedProxy( entityKey, asPersistentAttributeInterceptable( target ) );
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:219]: Is this change actually tested? I did not see anything obvious in the tests that would be creatiung temporary sessions. Are the test entities enhanced or something?

Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:219]: Though the PR doesn't involve bytecode enhancement we have tests that rely on this code. There was a failing test due to to the fact that we were adding the uninitialized bytecode proxy (along with an array of `UNFETCHED_PROPERTY`s as the state) as an actual entity here, marking the holder as `INITIALIZED`. To me this looked wrong, and `addEnhancedProxy()` seemed like the perfect method to use instead.

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:219]: You mean this test started failing after the other changes in the PR?  Just trying to understand :)

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementAsProxyLazinessInterceptor.java:219]: Ok, that makes sense.

**Pull request merged**